### PR TITLE
Disable operator cache for pull

### LINKED ISSUE #6284: Four table inner join drops rows unless joins are sequential
### What happens?

I have 15 tables with the same schema, to join on an integer column.  All tables have the same number of records and same values in the PK join column.  Joining any 3 works, but joining 4 or more drops a small percentage of the rows.  I join like so

```
   FROM T0 INNER JOIN T1 on T1.record_nb = T0.record_nb
           INNER JOIN T2 on T2.record_nb = T0.record_nb
   ...
           INNER JOIN T14 on T14.record_nb = T0.record_nb
```

Altering this to join each table on the previous table causes the SQL to succeed.  No rows are dropped.  IOW, this works:

```
   FROM T0 INNER JOIN T1 on T1.record_nb = T0.record_nb
           INNER JOIN T2 on T2.record_nb = T1.record_nb
   ...
           INNER JOIN T14 on T14.record_nb = T13.record_nb
```

but uses significantly more memory.

Any 3 tables join properly with either FROM clause, but any 4 or more yield inconsistent results.  Each execution yields a different number of rows.

### To Reproduce

This drops rows:
```python
import polars as pl
import duckdb
import random

record_nb = set(random.randint(1,9_999_999) for _ in range(2_000_000))

frames = [pl.DataFrame({'record_nb': list(record_nb), f'x_{i}': 0.0, f'y_{i}': 1.0}) 
          for i in range(4)]
tbls = [df.to_arrow() for df in frames]
with duckdb.connect() as con:
    select_ = 'SELECT T0.record_nb'
    from_ = ' FROM T0 '
    for i, (tbl, df) in enumerate((zip(tbls, frames))):
        con.register(f'T{i}', tbl)
        select_ += ',' + ','.join(f'\n    T{i}.{x} {x}' for x in (f'x_{i}', f'y_{i}') )
        if i > 0:
            from_ += f'\n   INNER JOIN T{i} on T0.record_nb = T{i}.record_nb'
    sql = f'{select_}\n{from_}'
    con.execute(sql)
    result_pd = con.fetchdf()

print(f'frames have shape {frames[0].shape} and joined result has shape {result_pd.shape}')
```
Since all frames have the same values in the column being joined, the result should have the same number of rows, but the output is:

`frames have shape (1812272, 3) and joined result has shape (1811160, 9)`

If the `with duckdb.connect() as con` block is run repeatedly, the number of rows in the result changes each time.

Reducing the dataframe sizes by 10x, or number of tables to 3 eliminates the discrepancy.  If I change `T0.record_nb` to `T{i-1}.record_nb` in the `INNER JOIN` expression, the problem disappears.

I tried using `.to_pandas()` in place of `.to_arrow()` but had the same result.

### OS:

Ubuntu 18.04 on x64

### DuckDB Version:

0.7.0

### DuckDB Client:

Python

### Full Name:

Kendall Bailey

### Affiliation:

Figure Technologies

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by samansmink**
fixes #6284

The fix is disabling the operator caching on streaming query results alltogether. This code was quite complex and we want to move away from the pull based streaming anyways to be able to support parallelizing streaming query results

### CODE REVIEW

File: src/execution/physical_operator.cpp [hunk 1]
```diff
@@ -252,15 +252,15 @@ OperatorResultType CachingPhysicalOperator::Execute(ExecutionContext &context, D
...
-
-		if (context.pipeline->GetSink() && context.pipeline->GetSink()->RequiresBatchIndex()) {
+		} else if (!context.pipeline->GetSink()) {
+			// Disabling for pipelines without Sink, i.e. when pulling
 			state.can_cache_chunk = false;
-		}
-
-		if (context.pipeline->IsOrderDependent()) {
+		} else if (context.pipeline->GetSink()->RequiresBatchIndex()) {
+			state.can_cache_chunk = false;
```
Comment (Mytherin) [inline src/execution/physical_operator.cpp:262]: Does `IsOrderDependent` not imply this condition as well?

File: test/api/test_api.cpp [hunk 2]
```diff
@@ -554,3 +554,43 @@ TEST_CASE("Issue #4583: Catch Insert/Update/Delete errors", "[api]") {
 	result = con.SendQuery("SELECT MIN(c0) FROM t0;");
 	REQUIRE(CHECK_COLUMN(result, 0, {1}));
 }
+
+TEST_CASE("Issue #6284: CachingPhysicalOperator in pull causes issues", "[api]") {
```
Comment (Mytherin) [inline test/api/test_api.cpp:558]: Slow test maybe?

Comment (samansmink) [reply test/api/test_api.cpp]: good point

File: src/execution/physical_operator.cpp (hunk 1 above)
Comment (samansmink) [reply src/execution/physical_operator.cpp:262]: decided to leave as is as there might be cases where its not at some point

**Pull request merged**